OperatorDef <==> NodeProto Conversion

houseroad opened the pull request:
Operator level proto conversion between (new) torch proto and (old) caffe2 proto.

BIT-silence commented on caffe2/proto/torch_pb.h:
```diff
@@ -0,0 +1,3 @@
+#pragma once
```
https://google.github.io/styleguide/cppguide.html#The__define_Guard

#ifndef CAFFE2_PROTO_TORCH_PB_H_
#define CAFFE2_PROTO_TORCH_PB_H_

...

#endif // CAFFE2_PROTO_TORCH_PB_H_

BIT-silence commented on caffe2/python/pybind_state.cc:
```diff
@@ -1527,6 +1534,46 @@ void addGlobalMethods(py::module& m) {
+    OperatorDefToNodeProto(op_def, &node);
+    std::string ret;
+    node.SerializeToString(&ret);
+    return ret;
```
return node.SerializeAsString();

https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.message_lite#MessageLite.SerializeAsString.details

BIT-silence commented on caffe2/python/pybind_state.cc:
```diff
@@ -1527,6 +1534,46 @@ void addGlobalMethods(py::module& m) {
+    NodeProtoToOperatorDef(node_proto, &op_def);
+    std::string ret;
+    op_def.SerializeToString(&ret);
+    return ret;
```
ditto.

BIT-silence commented on caffe2/utils/proto_convert.cc:
```diff
@@ -0,0 +1,190 @@
+    attr->set_name("control_input");
+    attr->set_type(::torch::AttributeProto_AttributeType::
+                       AttributeProto_AttributeType_STRINGS);
+    for (int i = 0; i < def.control_input_size(); ++i) {
```
attr->mutable_strings()->CopyFrom(def.control_input());
or
for (const string& input : def.control_input()) {
  attr->add_strings(input);
}

BIT-silence commented on caffe2/utils/proto_convert.cc:
```diff
@@ -0,0 +1,190 @@
+      CAFFE_ENFORCE(attr.has_s());
+      def->set_engine(attr.s());
+    } else if (attr.name() == "control_input") {
+      for (int j = 0; j < attr.strings_size(); ++j) {
```
ditto

BIT-silence commented on caffe2/utils/proto_convert.h:
```diff
@@ -0,0 +1,22 @@
+#pragma once
```
#ifndef CAFFE2_UTILS_PROTO_CONVERT_H_
#define CAFFE2_UTILS_PROTO_CONVERT_H_

#endif // CAFFE2_UTILS_PROTO_CONVERT_H_

xiaomengy reviewed (approved):
LGTM

dzhulgakov commented on caffe2/python/pybind_state.cc:
```diff
@@ -1186,11 +1187,17 @@ void addGlobalMethods(py::module& m) {
   });
   m.def("nets", []() { return gWorkspace->Nets(); });
-  m.def("run_operator_once", [](const py::bytes& op_def) {
+  m.def("run_operator_once", [](const py::bytes& op_def, bool legacy_proto=true) {
```
hm, adding it to this API feels a bit out of place now. Is it just for test?

houseroad replied:
Right now, it's just for test purpose.

houseroad committed:
OperatorDef <==> NodeProto Conversion (#11621)

Summary:
Operator level proto conversion between (new) torch proto and (old) caffe2 proto.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/11621

Reviewed By: BIT-silence

Differential Revision: D9892422

Pulled By: BIT-silence

fbshipit-source-id: f792a8c388aeec3382ee6b264b4101334d45dc36
